### PR TITLE
JNG-4250 Add PSM Archetype

### PR DESCRIPTION
**Pull request by robertcsakany**
<table>
<td>
  <a href="https://blackbelt.atlassian.net/browse/JNG-4250" title="JNG-4250" target="_blank">StoryJNG-4250</a>  Add PSM generator archetype
  </td></table>
  <br />

### CODE REVIEW

**Review by bencegelei — APPROVED** [review]

File: judo-psm-archetype-generator/pom.xml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                </configuration>
+            </plugin>
+
+        </plugins>
+    </build>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.eclipse.emf</groupId>
+            <artifactId>org.eclipse.emf.ecore</artifactId>
```
Comment (noherczeg) [inline judo-psm-archetype-generator/pom.xml]: Do we really need EMF stuff here?

File: judo-psm-archetype-generator/src/main/java/hu/blackbelt/judo/psm/archetype/generator/GeneratorTemplate.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.*;
+
+
+/**
+ * Generator template describes one generation properties.
+ * <ul>
+ *     <li>factoryExpression: A SpringEL Expression which can result a collection which will be used as list ofg generated output. The collection's elements is used
```
Comment (noherczeg) [inline judo-psm-archetype-generator/src/main/java/hu/blackbelt/judo/psm/archetype/generator/GeneratorTemplate.java]: "ofg"?

File: judo-psm-archetype-generator/src/test/java/hu/blackbelt/judo/psm/archetype/generator/Northwind2UiApplicationTest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.stream.StreamSupport;
+
+import static hu.blackbelt.judo.meta.psm.runtime.PsmModel.SaveArguments.psmSaveArgumentsBuilder;
+import static hu.blackbelt.judo.meta.psm.runtime.PsmModel.buildPsmModel;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.linesOf;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@Slf4j
+public class Northwind2UiApplicationTest {
```
Comment (noherczeg) [inline judo-psm-archetype-generator/src/test/java/hu/blackbelt/judo/psm/archetype/generator/Northwind2UiApplicationTest.java]: UI?

File: judo-psm-archetype-generator/src/test/resources/override1/test1/actorname.hbs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Name: {{ actorType.name }}
```
Comment (noherczeg) [inline judo-psm-archetype-generator/src/test/resources/override1/test1/actorname.hbs]: How can this work if some actor helper thingy was commented out in one of the files above?

File: judo-psm-archetype-maven-plugin/README.adoc [hunk 5]
```diff
@@ -0,0 +1,207 @@
+JUDO Tatami PSM Archetype maven plugin
+========================================
+
+This plugin manages and executes generators for JUDO PSM models.
+
+It generates text based codes / reports from PSM model with the given templates.
+
+## Requirements
+
+- Maven 3.6 and Java 11
```
Comment (noherczeg) [inline judo-psm-archetype-maven-plugin/README.adoc]: 3.6?

File: judo-psm-archetype-maven-plugin/README.adoc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+<6> Template parameters. The defined parameters here can be accessed in SpringEL and handlebars
+templates woth the same name.
+
+<7> ValueHelpers. It must implement `com.github.jknack.handlebars.ValueResolver`. It can be used to extend types with methods. These
+only used in templates, it cannot be used in SpringEL.
+
+## Generation
+
+For code generation handlebars templates is used. The `<project type>-project.yaml` file
```
Comment (noherczeg) [inline judo-psm-archetype-maven-plugin/README.adoc]: templates -> template.

File: judo-psm-archetype-maven-plugin/README.adoc [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+## Override templates in generation
+
+The template override can contain a `<project type>-project.yaml` which can be empty, on that case the
+existing templates can be overrided only. The content of project file is used to override existing
+template definition or can be added new templates. The ovirrided preferences reverse ordered, so the last defined override is the strongest.
+All of the templates can be decorated, when the original file name is suffixed with `override.hbs`. When it is
+defined the original one can be included with the standard fragment syntax of handlebars.
+
+### Example:
+To decorate `main.dart.hbs,` put `main.dart.override.hbs` file. The original content can be referenced:
```
Comment (noherczeg) [inline judo-psm-archetype-maven-plugin/README.adoc]: Dart? :)

File: judo-psm-archetype-maven-plugin/README.adoc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+```
+The original content:
+{{> flutter/lib/main.dart.hbs }}
```
Comment (noherczeg) [inline judo-psm-archetype-maven-plugin/README.adoc]: Flutter? :D

File: judo-psm-archetype-maven-plugin/pom.xml [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+		<dependency>
+			<groupId>org.apache.maven.plugin-tools</groupId>
+			<artifactId>maven-plugin-annotations</artifactId>
+			<version>3.4</version>
+			<scope>provided</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>io.swagger</groupId>
+			<artifactId>swagger-core</artifactId>
```
Comment (noherczeg) [inline judo-psm-archetype-maven-plugin/pom.xml]: Swagger?

**Review by noherczeg — CHANGES_REQUESTED** [review]

**Review by gaborflorian — APPROVED** [review]

File: judo-psm-archetype-generator/pom.xml (hunk 1 above)
Comment (robertcsakany) [reply judo-psm-archetype-generator/pom.xml]: Removed

File: judo-psm-archetype-generator/src/test/resources/override1/test1/actorname.hbs (hunk 4 above)
Comment (robertcsakany) [reply judo-psm-archetype-generator/src/test/resources/override1/test1/actorname.hbs]: For PSM model there is an implicit helper.

**Pull request merged**